DMTN-148.tex

```diff
@@ -50,9 +50,255 @@
+  image data or an explicit calibration class in the \verb|ip_isr|
+  package.\footnote{This will require class reorganization to migrate
+    the Defects and Curve calibrations definitions.}
+\item All calibration products should be constructed from a DM Task in
+  the \verb|cp_pipe| package, and have the input data and
+  configuration parameters persisted with the calibration itself.
+  This guarantees that any product can be reconstructed in the future.
+  Exceptions will exist for external or lab created calibrations,
+  although the majority of used calibrations will likely be produced
+  directly.
+\item These provenance, config, and final product files may be saved
+  to an \verb|obs_CAMERA_data| package in git for convenience.  This also allows
+  these files to be migrated to other Butler repositories in a
+  consistent fashion.
+\item All calibration products are accepted for use and have a date
+  range set manually, preventing test runs from being used
+  inadvertently.  Calibrations ingested from git should have sufficient
+  metadata that supply appropriate date ranges.
+\item Calibration information stored in a lsst.afw.cameraGeom object
+  may be used, but is always assumed to be overridden by supplied
+  calibration products.  No new calibration information should be
+  stored in these objects, and currently existing calibration
+  information should be deprecated.
+
+\end{itemize}
+
+\section{The Current State}
+
+\subsection{Standard Calibrations}
+
+The classic calibration products, \verb|BIAS|, \verb|DARK|,
+\verb|FLAT|, and \verb|FRINGE|, all write FITS files that are indexed
+in Butler's calibRegistry.sqlite3 repository database.  These files
+are created by \verb|pipe_drivers|/constructCalibs.py, and are
+ingested into the repository upon acceptance by
+\verb|pipe_tasks|/ingestCalibs.py.  This ingestion process sets the
+``validStart'' and ``validEnd'' values to restrict the usage of the
+calibration, as well as some identifying string which is used (along
+with calibration type, filter, and detector) to uniquely identify the
+calibration on disk.  We have been using the ``calibDate'', indicating
+when the calibration was constructed or ingested, as this string.  For
+calibrations that are generated by \verb|cp_pipe|, it may be better to
+use the Butler collection that the calibration was generated in (which
+will enforce uniqueness on simultaneous processing).  Calibrations
+that are ingested from git could use the repository SHA1 hash, which
+would provide similar uniqueness.
+
+Although not formally a ``classic'' calibration product, the \verb|SKY|
+calibration is produced, stored, and accessed the same way.
+
+These calibrations are required to have the following header keywords
+defined, which are used during ingest to define the rows of the
+appropriate calibration database table.  I believe only \verb|OBSTYPE|
+is required right now.
+
+\begin{tabular}{l l}
```

Comment:
Do we have a normative document defining these keywords? I'm not sure if this is capturing what just happens to be implemented in code, or if it's referring to another document where this is defined, or even if this table is supposed, in itself, to be normative.

Reply:
My understanding is that much of this comes out of what pipe_drivers is writing out what pipe_tasks ingestCalibs was requiring. I don't remember seeing CALIB_ID documented anywhere.

Reply:
That's one that I'm still unsure about, as it exists, but I can't see it being used anywhere.  My plan in gen3 is to simply dump the output dataset dimension here.